cpp/src/arrow/extension/tensor_array.cc

```diff
@@ -0,0 +1,103 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/extension/tensor_array.h"
+#include "arrow/tensor.h"
+
+namespace arrow {
+namespace extension {
+
+static std::string get_name(const std::shared_ptr<DataType>& type,
+                            const std::vector<int64_t>& shape,
+                            const std::vector<int64_t>& strides) {
+  std::stringstream s;
+  s << "arrow.extension.tensor<type=" << *type << ", shape=(";
+  for (uint64_t i = 0; i < shape.size(); i++) {
+    s << shape[i];
+    if (i < shape.size() - 1) {
+      s << ", ";
+    }
+  }
+  s << "), strides=(";
+  for (uint64_t i = 0; i < strides.size(); i++) {
+    s << strides[i];
+    if (i < strides.size() - 1) {
+      s << ", ";
+    }
+  }
+  s << ")>";
+  return s.str();
+}
+
+std::string TensorArrayType::extension_name() const {
+  return get_name(type(), shape(), strides());
+}
+
+const char* TensorArrayType::type_name() const {
+  std::string extension_name = get_name(type(), shape(), strides());
+  return extension_name.c_str();
+}
+
+std::shared_ptr<ExtensionType> tensor_array(const std::shared_ptr<DataType>& type,
+                                            const std::vector<int64_t>& shape,
+                                            const std::vector<int64_t>& strides) {
+  std::string type_name = get_name(type, shape, strides);
+  auto ext_type = GetExtensionType(type_name);
+  if (!ext_type) {
+    DCHECK_OK(
+        RegisterExtensionType(std::make_shared<TensorArrayType>(type, shape, strides)));
+  }
+  return GetExtensionType(type_name);
+}
+
+bool TensorArrayType::ExtensionEquals(const ExtensionType& other) const {
+  const auto& other_ext = static_cast<const TensorArrayType&>(other);
```

Comment:
You can't do that without first checking that `other` has the right extension type. This is what `extension_name()` is for.

Reply:
Added extension_name equality check.